### PR TITLE
P2P: Improve performance by not serializing and deserializing blocks from block log

### PR DESCRIPTION
**Pull request by linh2931**
Currently, when a peer asks for a block, if the block is retrieved from the block log, `net plugin`
- reads the block form the block log and deserializes it
- then serializes it before sending it to the peer

This can be expensive and add up to considerable CPU, see #611.

Those conversions are not necessary, as the serialization format in block log is the same as on P2P.

This PR adds `read_serialized_block_by_num()` to `block_log` and `fetch_serialized_block_by_number()` to `controller`, and updates `net plugin` to use the new interface.

Resolves https://github.com/AntelopeIO/spring/issues/612

### CODE REVIEW

File: libraries/chain/block_log.cpp [hunk 1]
```diff
@@ -615,6 +628,68 @@ namespace eosio { namespace chain {
...
+                  // Just fall back to the regular `retry_read_block_by_num` and then serialize.
+                  auto ret = retry_read_block_by_num(block_num);
+                  if (ret) {
+                     return fc::raw::pack(*ret);
+                  } else {
+                     return {};
+                  }
+               }
+
+               if (verify_block_num == verify_block_num_t::yes) {
```
Comment (linh2931) [inline libraries/chain/block_log.cpp:646]: Or should we always do a full deserialization here to make sure the block is valid but still return the serialized format? This matches the validation behavior of the original `read_block_by_num()` but still saves Net Plugin from doing serialization.

**Comment by ericpassmore** [comment]
Note:start
category: Other
component: Internal
summary: Improve P2P performance by maintaining serialized format from block log to peer syncing.
Note:end

File: libraries/chain/block_log.cpp [hunk 2]
```diff
@@ -197,6 +197,23 @@ namespace eosio { namespace chain {
...
+      uint32_t get_block_num_at(fc::datastream<fc::cfile>& file, uint64_t position) {
+         // to derive blknum_offset==14 see block_header.hpp and note on disk struct is packed
+         //   block_timestamp_type timestamp;                  //bytes 0:3
+         //   account_name         producer;                   //bytes 4:11
+         //   uint16_t             confirmed;                  //bytes 12:13
+         //   block_id_type        previous;                   //bytes 14:45, low 4 bytes is big endian block number
+         //   of previous block
+
+         file.seek_end(0);
+         EOS_ASSERT(position <= file.tellp(), block_log_exception, "Invalid block position ${position}",
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:209]: This really should be:
```suggestion
         EOS_ASSERT(position +  blknum_offset + sizeof(uint32_t) <= file.tellp(), block_log_exception, "Invalid block position ${position}",
```

with `blknum_offset` declared before.

File: libraries/chain/block_log.cpp [hunk 3]
```diff
@@ -474,6 +479,12 @@ namespace eosio { namespace chain {
          };
          std::optional<signed_block_with_id> head;
 
+         enum class verify_block_num_t {
+            yes, // Verify block number when read a serialized block.
+            no   // Does not verify block number when read a serialized block.
+                 // This is for the case where block id is vefified seperately.
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:485]:
```suggestion
            yes, // Verify the block number when reading a serialized block.
            no   // Do not verify the block number when reading a serialized block.
                 // This is for the case where the block id is verified separately.
```

File: libraries/chain/block_log.cpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/chain/block_log.cpp:646]: Wouldn't a block in the block log be guaranteed to be valid?

File: libraries/chain/block_log.cpp [hunk 4]
```diff
@@ -1230,6 +1305,32 @@ namespace eosio { namespace chain {
...
+   }
+
+   std::vector<char> block_log::read_serialized_block_by_id(const block_id_type& id) const {
+      std::lock_guard g(my->mtx);
+
+      auto block_num = block_header::num_from_id(id);
+
+      std::optional<signed_block_header> h = my->read_block_header_by_num(block_num);
+      if (h) {
+         auto returd_id = h->calculate_id();
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:1320]:
```suggestion
         auto read_id = h->calculate_id();
```

File: libraries/chain/block_log.cpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/chain/block_log.cpp:646]: I don't understand whay we need this `verify_block_num`. I would think that the blocks in the block log are correct, and that therefore the block at position ` get_block_pos(block_num)` is guaranteed to have the block number `block_num`. Is that not the case?

File: libraries/chain/block_log.cpp [hunk 5]
```diff
@@ -615,6 +628,68 @@ namespace eosio { namespace chain {
...
+                  // current block is not the last block in the log file.
+                  uint64_t next_block_pos = get_block_pos(block_num + 1);
+                  if (next_block_pos == block_log::npos) {
+                     wlog("no position found for block ${n}", ("n", block_num + 1));
+                     return {};
+                  }
+                  if (next_block_pos <= pos) {
+                     wlog("next block position ${nn} should be greater than current block position ${cn}", ("nn", next_block_pos)("cn", pos));
+                     return {};
+                  }
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:669]: I think this can be `assert(next_block_pos > pos);`

File: libraries/chain/block_log.cpp [hunk 6]
```diff
@@ -615,6 +628,68 @@ namespace eosio { namespace chain {
...
+               uint32_t head_block_num = block_header::num_from_id(head->id);
+               constexpr uint32_t block_pos_size = sizeof(uint64_t); // size of block position field in the block log file
+
+               if (block_num < head_block_num) {
+                  // current block is not the last block in the log file.
+                  uint64_t next_block_pos = get_block_pos(block_num + 1);
+                  if (next_block_pos == block_log::npos) {
+                     wlog("no position found for block ${n}", ("n", block_num + 1));
+                     return {};
+                  }
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:665]: I think this can be: `assert(next_block_pos != block_log::npos);`

File: libraries/chain/block_log.cpp (hunk 1 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:646]: The original `read_block_by_num` checks `block_num`. I am little paranoid. Will discuss in design meeting.

File: libraries/chain/block_log.cpp (hunk 2 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:209]: This EOS_ASSERT was in the original function. Will correct it. Thanks.

File: libraries/chain/block_log.cpp (hunk 3 above)
Comment (greg7mdp) [reply libraries/chain/block_log.cpp]: still some typos remain.

Comment (linh2931) [reply libraries/chain/block_log.cpp]: This section of code was removed.

File: libraries/chain/block_log.cpp (hunk 5 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp]: It was decided to use EOS_ASSERT for preventing undefined behavior on our side and for consistency.

File: libraries/chain/block_log.cpp (hunk 6 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp]: It was decided to use EOS_ASSERT.

File: libraries/chain/block_log.cpp [hunk 7]
```diff
@@ -615,6 +628,68 @@ namespace eosio { namespace chain {
...
+               }
+
+               if (verify_block_num == verify_block_num_t::yes) {
+                  auto bnum = get_block_num_at(block_file, pos);
+                  if (block_num != bnum) {
+                     wlog("Wrong block num was read from block log. expected block_num: ${n}, returned block num: ${bnum}",
+                           ("n", block_num)("bnum", bnum));
+                     return {};
+                  }
+               }
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/block_log.cpp]: This code was removed.

File: libraries/chain/controller.cpp [hunk 8]
```diff
@@ -5514,6 +5514,13 @@ signed_block_ptr controller::fetch_block_by_id( const block_id_type& id )const {
    return signed_block_ptr();
 }
 
+std::vector<char> controller::fetch_serialized_block_by_id( const block_id_type& id )const { try {
```
Comment (heifner) [inline libraries/chain/controller.cpp:5517]: We don't need the `fetch_serialized_block_by_id`. If we do in the future, I think it should verify the retrieved serialized block is the correct block id instead of just assuming the block_num of the id will retrieve the correct block.

Comment (heifner) [reply libraries/chain/controller.cpp:5517]: See new PR https://github.com/AntelopeIO/spring/pull/803 that removes its usage.

**Review by heifner — CHANGES_REQUESTED** [review]
Still reviewing, but thought I would get these comments out so you know why PR #803 was created.

File: plugins/net_plugin/net_plugin.cpp [hunk 9]
```diff
@@ -1838,6 +1838,25 @@ namespace eosio {
...
+         const uint32_t payload_size = which_size + v.size();
+
+         const char* const header = reinterpret_cast<const char* const>(&payload_size); // avoid variable size encoding of uint32_t
+         const size_t buffer_size = message_header_size + payload_size;
+
+         auto send_buffer = std::make_shared<vector<char>>( buffer_size );
+         fc::datastream<char*> ds( send_buffer->data(), buffer_size );
+         ds.write( header, message_header_size );
+         fc::raw::pack( ds, unsigned_int( signed_block_which ) );
+         ds.write( v.data(), v.size() );
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1855]: Instead of `vector<char> read_serialized_block_by_num(block_num)` how about `void read_serialized_block_by_num(block_num, vector<char>& data)`. 

`read_serialized_block_by_num` would append to the `data` buffer. We can then avoid the copy here.

**Review by heifner — CHANGES_REQUESTED** [review]
Still reviewing, but thought I would get these comments out so you know why PR #803 was created.

File: libraries/chain/block_log.cpp [hunk 10]
```diff
@@ -245,10 +246,13 @@ namespace eosio { namespace chain {
             //   block_id_type        previous;                   //bytes 14:45, low 4 bytes is big endian block number
             //   of previous block
 
-            EOS_ASSERT(position <= size(), block_log_exception, "Invalid block position ${position}",
-                       ("position", position));
+            file.seek_end(0);
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:249]: Why did you change this to not use `size()`, and do a `seek_end()` instead?

File: libraries/chain/block_log.cpp [hunk 11]
```diff
@@ -615,6 +621,63 @@ namespace eosio { namespace chain {
...
+                  auto ret = retry_read_block_by_num(block_num);
+                  if (ret) {
+                     return fc::raw::pack(*ret);
+                  } else {
+                     return {};
+                  }
+               }
+
+               uint32_t head_block_num = block_header::num_from_id(head->id);
+               EOS_ASSERT(block_num <= head_block_num, block_log_exception,
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:640]: Doesn't this assert change the previous behavior of  `connection::blk_send`? 

Before, it would: 
`peer_ilog( this, "fetch block by id returned null, id ${id}", ("id", blkid) );`

Now it will:
`peer_elog( this, "caught assert on fetch_serialized_block_by_id, ${ex}, id ${id}", ("ex", ex.to_string())("id", blkid) );`

I don't think Areg said we had to always use `EOS_ASSERT`, but we had to avoid undefined behavior. In this case having an info message stating that the requested block was not found seems fine.

File: libraries/chain/block_log.cpp [hunk 12]
```diff
@@ -615,6 +621,63 @@ namespace eosio { namespace chain {
...
+
+               uint64_t block_size     = 0;
+               constexpr uint32_t block_pos_size = sizeof(uint64_t); // size of block position field in the block log file
+
+               if (block_num < head_block_num) {
+                  // current block is not the last block in the log file.
+                  uint64_t next_block_pos = get_block_pos(block_num + 1);
+
+                  EOS_ASSERT(next_block_pos != block_log::npos, block_log_exception,
+                             "no position found for block ${n}", ("n", block_num + 1));
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:652]: I think the next assert at line 653 is sufficient, and this one is not needed.

File: libraries/chain/block_log.cpp [hunk 13]
```diff
@@ -615,6 +621,63 @@ namespace eosio { namespace chain {
...
+                  EOS_ASSERT(next_block_pos > pos + block_pos_size, block_log_exception,
+                             "next block position ${np} should be greater than current block position ${p} plus block position field size ${bps}",
+                             ("np", next_block_pos)("p", pos)("bps", block_pos_size));
+
+                  block_size = next_block_pos - pos - block_pos_size;
+               } else {
+                  // current block is the last block in the file.
+                  assert (block_num == head_block_num);
+
+                  auto file_size = std::filesystem::file_size(block_file.get_file_path());
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:662]: why not use `size()`

File: libraries/chain/block_log.cpp (hunk 10 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp]: It was a leftover from last move around. Reverted back.

File: libraries/chain/block_log.cpp (hunk 13 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp]: `basic_block_log` does not have `size()` to return the size of its block log file.

File: libraries/chain/block_log.cpp [hunk 14]
```diff
@@ -185,6 +185,16 @@ namespace eosio { namespace chain {
          return block;
       }
 
+      template <typename Stream>
+      void read_serialized_block(Stream&& ds, uint64_t block_size, std::vector<char>& buff) {
+         assert(buff.empty());
```
Comment (heifner) [inline libraries/chain/block_log.cpp:190]: The point of taking the vector by reference was to append to the already filled out message header from net_plugin. However, I realize now that the message_header includes the payload size which you don't know until here. 

We could let this method create the exact net_message needed. Call it `vector<char> create_net_message_from_block(block_num_type block_num)` then we could avoid the copy. 

If we don't want to add that knowledge to `block_log.cpp` then I guess we will have to pay for an extra copy of the `vector<char>` or do something more fancy and use two `boost::asio::const_buffer`s, one for the header and one for the payload.

File: libraries/chain/block_log.cpp [hunk 15]
```diff
@@ -185,6 +185,16 @@ namespace eosio { namespace chain {
...
+
+         try {
+            buff.resize(block_size);
+            ds.read(buff.data(), block_size); // append to buff
+         } catch (...) { buff.clear(); } // restore buff to its original state
```
Comment (heifner) [inline libraries/chain/block_log.cpp:195]: With this `catch` and the assert might as well just return the vector as it is the same thing with return value optimization.

File: libraries/chain/block_log.cpp [hunk 16]
```diff
@@ -603,6 +619,43 @@ namespace eosio { namespace chain {
             return pos;
          }
 
+         block_pos_size_t get_block_position_and_size(uint32_t block_num) {
```
Comment (heifner) [inline libraries/chain/block_log.cpp:622]: For this and the other added methods can you add unittests for them.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/block_log.cpp (hunk 14 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:190]: I was trying to insert the serialized bytes into the Net message but gave up due to complexities. I like the original returning vector approach: it is cleaner, no precondition assumption, and no knowledge of net plugin.

File: libraries/chain/block_log.cpp (hunk 15 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:195]: Will revert back to return vector.

File: libraries/chain/block_log.cpp (hunk 16 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:622]: Can the tests be in a separate PR. I plan to write some comprehensive tests.

Comment (heifner) [reply libraries/chain/block_log.cpp:622]: There is no rush to get this in, but doesn't matter to me.

Comment (linh2931) [reply libraries/chain/block_log.cpp:622]: https://github.com/AntelopeIO/spring/issues/808

File: libraries/chain/block_log.cpp [hunk 17]
```diff
@@ -615,6 +668,18 @@ namespace eosio { namespace chain {
             FC_LOG_AND_RETHROW()
          }
 
+         std::vector<char> read_serialized_block_by_num(uint32_t block_num) final {
+            try {
+               block_pos_size_t pos_size = get_block_position_and_size(block_num);
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:673]: I feel it would be cleaner with the following (and also `get_block_position_and_size` could return a `std::pair`)
```suggestion
               auto [position, size] = get_block_position_and_size(block_num);
```

File: libraries/chain/block_log.cpp [hunk 18]
```diff
@@ -603,6 +619,43 @@ namespace eosio { namespace chain {
...
+         block_pos_size_t get_block_position_and_size(uint32_t block_num) {
+            uint64_t pos = get_block_pos(block_num);
+
+            if (pos == block_log::npos) {
+               return block_pos_size_t {.position = block_log::npos, .size = 0};
+            }
+
+            assert(head);
+            uint32_t last_block_num = block_header::num_from_id(head->id);
+            assert(block_num <= last_block_num);
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/block_log.cpp]: I was debating about this. We have already found the block position for `block_num` so it should be in "good" shape. But as `block_num` comes from Net Plugin and to be paranoid, use EOS_ASSERT just in case.

File: libraries/chain/block_log.cpp (hunk 17 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:676]: I think struct provides stronger type check, especially when those two members are of the same type, and it is more clear in return statement `return block_pos_size_t {.position = pos, .size = block_size}` than just return `{pos, block_size}`.

Comment (linh2931) [reply libraries/chain/block_log.cpp]: I can still use structure binding to make it cleaner. Thanks.

File: libraries/chain/include/eosio/chain/log_catalog.hpp [hunk 19]
```diff
@@ -159,6 +164,50 @@ struct log_catalog {
...
+         if (block_num < first_block_num())
+            return {};
+
+         auto it = --collection.upper_bound(block_num);
+
+         if (block_num <= it->second.last_block_num) {
+            auto name = it->second.filename_base;
+            log_data.open(name.replace_extension("log"));
+            log_index.open(name.replace_extension("index"));
+            active_index = std::distance(collection.begin(), it);
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/log_catalog.hpp:201]: This is all the same code as `get_block_position`. Seems you could call `get_block_position` from `block_position_size` and not have all this duplicate code.

File: plugins/net_plugin/net_plugin.cpp [hunk 20]
```diff
@@ -1898,8 +1910,7 @@ namespace eosio {
    }
 
    // called from connection strand
-   size_t connection::enqueue_block( const signed_block_ptr& b, bool to_sync_queue) {
-      peer_dlog( this, "enqueue block ${num}", ("num", b->block_num()) );
+   size_t connection::enqueue_block( const std::vector<char>& b, bool to_sync_queue) {
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1913]: Pass `block_num` into `enqueue_block` so you can add back in the `peer_dlog`.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/include/eosio/chain/log_catalog.hpp (hunk 19 above)
Comment (linh2931) [reply libraries/chain/include/eosio/chain/log_catalog.hpp]: Thanks! Much simpler.

**Review by heifner — APPROVED** [review]

File: libraries/chain/block_log.cpp [hunk 21]
```diff
@@ -603,6 +619,46 @@ namespace eosio { namespace chain {
             return pos;
          }
 
+         block_pos_size_t get_block_position_and_size(uint32_t block_num) {
```
Comment (greg7mdp) [inline libraries/chain/block_log.cpp:622]: minor detail, I think it might be cleaner to declare the return value in the beginning and use it instead of intermadiate variables, as below, but no need to change:

```c++
         block_pos_size_t get_block_position_and_size(uint32_t block_num) {
            block_pos_size_t bps { .position = get_block_pos(block_num) };

            if (bps.position == block_log::npos)
               return bps;

            assert(head);
            uint32_t last_block_num = block_header::num_from_id(head->id);
            EOS_ASSERT(block_num <= last_block_num, block_log_exception,
                       "block_num ${bn} should not be greater than last_block_num ${lbn}",
                       ("bn", block_num)("lbn", last_block_num));

            constexpr uint32_t block_pos_size = sizeof(uint64_t); // size of block position field in the block log file

            if (block_num < last_block_num) {
               // current block is not the last block in the log file.
               uint64_t next_block_pos = get_block_pos(block_num + 1);

               EOS_ASSERT(next_block_pos > bps.position + block_pos_size, block_log_exception,
                          "next block position ${np} should be greater than current block position ${p} plus block position field size ${bps}",
                          ("np", next_block_pos)("p", bps.position)("bps", block_pos_size));

               bps.size = next_block_pos - bps.position - block_pos_size;
            } else {
               // current block is the last block in the file.

               block_file.seek_end(0);
               auto file_size = block_file.tellp();
               EOS_ASSERT(file_size > bps.position + block_pos_size, block_log_exception,
                          "block log file size ${fs} should be greater than current block position ${p} plus block position field size ${bps}",
                          ("fs", file_size)("p", bps.position)("bps", block_pos_size));

               bps.size = file_size - bps.position - block_pos_size;
            }

            return bps;
         }
```

File: libraries/chain/include/eosio/chain/log_catalog.hpp [hunk 22]
```diff
@@ -159,6 +164,26 @@ struct log_catalog {
       }
    }
 
+   std::optional<block_pos_size_t> get_block_position_and_size(uint32_t block_num) {
+      std::optional<uint64_t> pos = get_block_position(block_num);
+
+      if (!pos) {
+         return {};
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/include/eosio/chain/log_catalog.hpp:171]: Was trying to follow existing method patterns in the two files: `get_block_position` in `log_catalog.hpp` while `get_block_pos` in `block_log.cpp` returns `block_log::npos`.

File: libraries/chain/block_log.cpp (hunk 21 above)
Comment (linh2931) [reply libraries/chain/block_log.cpp:622]: I think constructing return value right before returning makes it easier to see what to be returned.

**Review by greg7mdp — APPROVED** [review]
